build/gen_stub.php

```diff
@@ -1344,70 +1417,497 @@ public function __clone()
     }
 }
 
-function initializeZval(string $zvalName, $value): string
+class EvaluatedValue
 {
-    $code = "\tzval $zvalName;\n";
+    /** @var mixed */
+    public $value;
 
-    switch (gettype($value)) {
-        case "NULL":
-            $code .= "\tZVAL_NULL(&$zvalName);\n";
-            break;
+    /** @var string|null */
+    public $type;
+
+    /** @var string|null */
+    public $cConstName;
 
-        case "boolean":
-            $code .= "\tZVAL_BOOL(&$zvalName, " . ((int) $value) . ");\n";
-            break;
+    /** @var bool */
+    public $isUnknownConstValue;
 
-        case "integer":
-            $code .= "\tZVAL_LONG(&$zvalName, $value);\n";
-            break;
+    /** @var ConstInfo|null */
+    public $originatingConst;
 
-        case "double":
-            $code .= "\tZVAL_DOUBLE(&$zvalName, $value);\n";
-            break;
+    /**
+     * @param iterable<ConstInfo> $allConstInfos
+     */
+    public static function createFromExpression(Expr $expr, ?string $cConstName, iterable $allConstInfos): EvaluatedValue
+    {
+        $originatingConst = null;
+        $isUnknownConstValue = null;
 
-        case "string":
-            if ($value === "") {
-                $code .= "\tZVAL_EMPTY_STRING(&$zvalName);\n";
-            } else {
-                $strValue = addslashes($value);
-                $code .= "\tzend_string *{$zvalName}_str = zend_string_init(\"$strValue\", sizeof(\"$strValue\") - 1, 1);\n";
-                $code .= "\tZVAL_STR(&$zvalName, {$zvalName}_str);\n";
+        $evaluator = new ConstExprEvaluator(
+            function (Expr $expr) use ($allConstInfos, &$cConstName, &$originatingConst, &$isUnknownConstValue) {
+                if ($expr instanceof Expr\ConstFetch || $expr instanceof Expr\ClassConstFetch) {
+                    if ($expr instanceof Expr\ClassConstFetch) {
+                        $constName = $expr->class->toString() . "::" . $expr->name->toString();
+                    } else {
+                        $constName = $expr->name->toString();
+                    }
+
+                    if ($constName !== "UNKNOWN") {
+                        foreach ($allConstInfos as $const) {
+                            if ($const->name->__toString() === $constName) {
+                                if ($cConstName === null) {
```

Comment:
Similarly

```php
const FOO = UNKNOWN;
const BAR = 'bar';
const BAZ = FOO . BAR;
```

```c
REGISTER_STRING_CONSTANT("BAR", "bar", CONST_CS | CONST_PERSISTENT);
REGISTER_STRING_CONSTANT("BAZ", "bar", CONST_CS | CONST_PERSISTENT);
```

I think any expression containing a `UNKNOWN` value should be converted to `UNKNOWN` itself.

Btw sorry if these are edge cases that are not worth covering, it's just things I've noticed.

Reply:
> What if multiple constants are referenced in a single expressions, like FOO . BAR? This would then just point to the first one. I'm not sure what the implication of that is.

As I mentioned above, this is a difficult issue to tackle, as it requires PHP expressions to be rebuilt for C by swapping PHP constants with C constants + handling special cases. E.g. in case of expressions containing string concatenations, it would even need multiple statements in C. So I don't think string concats should be supported, possibly among a few other operations.

> Class constants referencing inexistent class constants are simply ignored:

Yes, this is a deliberate choice in order to be able to add class constants with `UNKNOWN` values. This is important because this way class synopses can list these constants and class entry generation won't be affected.

Reply:
Ah, just saw that you wrote the same:

> Same for class constants. For constants, it should error or be ignored. For class constants it should just be ignored since it's still useful for class synopses.

Yes, you are right!